test/hotspot/jtreg/serviceability/jvmti/vthread/TestPinCaseWithCFLH/TestPinCaseWithCFLH.java

```diff
@@ -0,0 +1,76 @@
+/*
+ * Copyright (c) 2024, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+import java.lang.instrument.ClassFileTransformer;
+import java.lang.instrument.IllegalClassFormatException;
+import java.lang.instrument.Instrumentation;
+import java.security.ProtectionDomain;
+import jdk.test.lib.thread.VThreadPinner;
+
+/*
+ * @test
+ * @summary javaagent + tracePinnedThreads will cause jvm crash/ run into deadlock when the virtual thread is pinned
+ * @library /test/lib
+ * @requires vm.continuations
+ * @requires vm.jvmti
+ * @modules java.base/java.lang:+open
+ * @compile TestPinCaseWithCFLH.java
+ * @run driver jdk.test.lib.util.JavaAgentBuilder
```

Comment:
Hello @jia-wei-tang, does adding an additional `@build jdk.test.lib.thread.VThreadPinner` line before the `@run` line help fix the `NoClassDefFoundError` you note in your comment?

Reply:
Hello @jia-wei-tang, I see that a `@build jdk.test.lib.util.JavaAgentBuilder` was added in the latest update to the PR. Did that help solve the NoClassDefFoundError you were running into when running those tests? I find it surprising that this specific class is required in the `@build` declaration, since that class itself is part of the `@run` directive.

Reply:
> Hello @jia-wei-tang, I see that a `@build jdk.test.lib.util.JavaAgentBuilder` was added in the latest update to the PR. Did that help solve the NoClassDefFoundError you were running into when running those tests? I find it surprising that this specific class is required in the `@build` declaration, since that class itself is part of the `@run` directive.

It cannot solve the NoClassDefFoundError. Should I try to add `@build jdk.test.lib.thread.VThreadPinner`?

Reply:
> Should I try to add @build jdk.test.lib.thread.VThreadPinner?

Yes, please give that a try.

Reply:
It cannot solve the NoClassDefFoundError. Besides, this new testcase is not included in linux-x86 / test (hs/tier1 serviceability).

Reply:
I see that the `NoClassDefFoundError` failure that you are mentioning is actually being reported even in the GitHub Actions job failures. It does look odd (although might be a pre-known jtreg issue). It will need a bit of investigation to see what's going on.

Reply:
I was able to reproduce this `NoClassDefFoundError` in this new test locally. I will take a look to see if I can figure out what's going on.

Reply:
I looked into this locally and this is a (known) bug in jtreg https://bugs.openjdk.org/browse/CODETOOLS-7902847. 

What's happening here is that the tests are launched using `make test TEST=test/hotspot/jtreg/:tier1_serviceability`. One of those tests is the (pre-existing unrelated to this PR) `AgentWithVThreadTest`. That test has a `@compile AgentWithVThread.java AgentWithVThreadTest.java`. This triggers compilation of those classes and also any referenced classes in those 2 classes. One such class happens to be the test library's `jdk.test.lib.Utils`. This is a test library class (used/referenced indirectly in that test). jtreg ends up issuing a `javac` command with destination directory as the AgentWithVThreadTest's test specific work directory:

Reply:
I looked into this locally and this is a (known) bug in jtreg https://bugs.openjdk.org/browse/CODETOOLS-7902847.

What's happening here is that the tests are launched using make test TEST=test/hotspot/jtreg/:tier1_serviceability. One of those tests is the (pre-existing unrelated to this PR) AgentWithVThreadTest. That test has a `@compile AgentWithVThread.java AgentWithVThreadTest.java`. This triggers compilation of those classes and also any referenced classes in those 2 classes. One such class happens to be the test library's jdk.test.lib.Utils. This is a test library class (used/referenced indirectly in that test). jtreg ends up issuing a javac command with destination directory as the AgentWithVThreadTest's test specific work directory:

```
-d build/macosx-aarch64/test-support/jtreg_test_hotspot_jtreg_tier1_serviceability/classes/0/serviceability/jvmti/vthread/premain/AgentWithVThreadTest.d

```
So the `jdk.test.lib.Utils.class` (along with other classes) file ends up being compiled to `build/macosx-aarch64/test-support/jtreg_test_hotspot_jtreg_tier1_serviceability/classes/0/serviceability/jvmti/vthread/premain/AgentWithVThreadTest.d/jdk/test/lib/Utils.class`. During this compilation, the `jdk.test.lib.util.JavaAgentBuilder` doesn't get compiled because that class isn't referenced by AgentWithVThreadTest (neither directly or indirectly).

Then during the same test execution, jtreg notices a `@run` statement:

```
@run driver jdk.test.lib.util.JavaAgentBuilder ....
```

And to launch that run action, it first builds and compiles the jdk.test.lib.util.JavaAgentBuilder and since this is a test library class, jtreg ends up launching `javac` with a destination directory which is common/shared by multiple tests:

```
-d build/macosx-aarch64/test-support/jtreg_test_hotspot_jtreg_tier1_serviceability/classes/0/test/lib
```

Additionally, since this is being compiled in context of the AgentWithVThreadTest, jtreg also passes the test specific work directory (`build/macosx-aarch64/test-support/jtreg_test_hotspot_jtreg_tier1_serviceability/classes/0/serviceability/jvmti/vthread/premain/AgentWithVThreadTest.d`) as the classpath to the javac command. So effectively, this compilation ends up finding the `jdk.test.lib.Utils.class` in the test specific directory and doesn't recompile to the shared location. Since the `jdk.test.lib.util.JavaAgentBuilder` hasn't yet been compiled nor is located in the test specific work directory, javac ends up compiling it and placing it in the destination directory which was passed to the javac invocation and happens to be a shared directory for tests `build/macosx-aarch64/test-support/jtreg_test_hotspot_jtreg_tier1_serviceability/classes/0/test/lib`.

Effectively we now have a test library class (the `JavaAgentBuilder`) in the common shared directory and some of its referenced classes (`Utils.class`) in a test specific directory.

At a later point in time, this new test `TestPinCaseWithCFLH` being proposed in this PR, gets launched and jtreg notices the `@run` statement:

```
@run driver jdk.test.lib.util.JavaAgentBuilder ...
```

Just like previously, since this is a test library class, jtreg tries to locate this class in the shared directory and it finds it in the shared directory `build/macosx-aarch64/test-support/jtreg_test_hotspot_jtreg_tier1_serviceability/classes/0/test/lib` (since it was compiled to that location by an unrelated test). Since it finds the class, jtreg then skips compilation of that test library class and thus the `Utils.class` (or any of the classes referenced by `JavaAgentBuilder`) isn't recompiled again. jtreg then launches the test with this shared directory and this test's specific work directory (which is different from the AgentWithVThreadTest's work directory) in the classpath. Since neither of these directories contain the `Utils.class`, we end up with this missing class error.

Like I noted, this is a known problem in jtreg. I'll see if we can do something in that area.

For now though, I think one workaround is to add a:

```
@clean jdk.test.lib.util.JavaAgentBuilder
```

before the `@run` tag to allow for the jdk.test.lib.util.JavaAgentBuilder and its referenced classes to be compiled afresh (into the shared test library directory). This is what the change would look like in your PR:

```diff
diff --git a/test/hotspot/jtreg/serviceability/jvmti/vthread/TestPinCaseWithCFLH/TestPinCaseWithCFLH.java b/test/hotspot/jtreg/serviceability/jvmti/vthread/TestPinCaseWithCFLH/TestPinCaseWithCFLH.java
index 02755a0289f..60564115f51 100644
--- a/test/hotspot/jtreg/serviceability/jvmti/vthread/TestPinCaseWithCFLH/TestPinCaseWithCFLH.java
+++ b/test/hotspot/jtreg/serviceability/jvmti/vthread/TestPinCaseWithCFLH/TestPinCaseWithCFLH.java
@@ -34,7 +34,7 @@
  * @requires vm.jvmti
  * @modules java.base/java.lang:+open
  * @compile TestPinCaseWithCFLH.java
- * @build jdk.test.lib.Utils
+ * @clean jdk.test.lib.util.JavaAgentBuilder
  * @run driver jdk.test.lib.util.JavaAgentBuilder
  *             TestPinCaseWithCFLH TestPinCaseWithCFLH.jar
  * @run main/othervm/timeout=100  -Djdk.virtualThreadScheduler.maxPoolSize=1

```

I admit it's odd to be expecting the test to be doing this, but I think this should make the test stable. Can you give it a try?

Reply:
I was inspired by [CODETOOLS-7901986](https://bugs.openjdk.org/browse/CODETOOLS-7901986). By adding `@build jdk.test.lib.Utils` all tests are passed. The GitHub Actions jobs are finished successfully, too. I hope I could get a final review now.

Reply:
> I looked into this locally and this is a (known) bug in jtreg https://bugs.openjdk.org/browse/CODETOOLS-7902847.
> 
> What's happening here is that the tests are launched using make test TEST=test/hotspot/jtreg/:tier1_serviceability. One of those tests is the (pre-existing unrelated to this PR) AgentWithVThreadTest. That test has a `@compile AgentWithVThread.java AgentWithVThreadTest.java`. This triggers compilation of those classes and also any referenced classes in those 2 classes. One such class happens to be the test library's jdk.test.lib.Utils. This is a test library class (used/referenced indirectly in that test). jtreg ends up issuing a javac command with destination directory as the AgentWithVThreadTest's test specific work directory:
> 
> ```
> -d build/macosx-aarch64/test-support/jtreg_test_hotspot_jtreg_tier1_serviceability/classes/0/serviceability/jvmti/vthread/premain/AgentWithVThreadTest.d
> ```
> 
> So the `jdk.test.lib.Utils.class` (along with other classes) file ends up being compiled to `build/macosx-aarch64/test-support/jtreg_test_hotspot_jtreg_tier1_serviceability/classes/0/serviceability/jvmti/vthread/premain/AgentWithVThreadTest.d/jdk/test/lib/Utils.class`. During this compilation, the `jdk.test.lib.util.JavaAgentBuilder` doesn't get compiled because that class isn't referenced by AgentWithVThreadTest (neither directly or indirectly).
> 
> Then during the same test execution, jtreg notices a `@run` statement:
> 
> ```
> @run driver jdk.test.lib.util.JavaAgentBuilder ....
> ```
> 
> And to launch that run action, it first builds and compiles the jdk.test.lib.util.JavaAgentBuilder and since this is a test library class, jtreg ends up launching `javac` with a destination directory which is common/shared by multiple tests:
> 
> ```
> -d build/macosx-aarch64/test-support/jtreg_test_hotspot_jtreg_tier1_serviceability/classes/0/test/lib
> ```
> 
> Additionally, since this is being compiled in context of the AgentWithVThreadTest, jtreg also passes the test specific work directory (`build/macosx-aarch64/test-support/jtreg_test_hotspot_jtreg_tier1_serviceability/classes/0/serviceability/jvmti/vthread/premain/AgentWithVThreadTest.d`) as the classpath to the javac command. So effectively, this compilation ends up finding the `jdk.test.lib.Utils.class` in the test specific directory and doesn't recompile to the shared location. Since the `jdk.test.lib.util.JavaAgentBuilder` hasn't yet been compiled nor is located in the test specific work directory, javac ends up compiling it and placing it in the destination directory which was passed to the javac invocation and happens to be a shared directory for tests `build/macosx-aarch64/test-support/jtreg_test_hotspot_jtreg_tier1_serviceability/classes/0/test/lib`.
> 
> Effectively we now have a test library class (the `JavaAgentBuilder`) in the common shared directory and some of its referenced classes (`Utils.class`) in a test specific directory.
> 
> At a later point in time, this new test `TestPinCaseWithCFLH` being proposed in this PR, gets launched and jtreg notices the `@run` statement:
> 
> ```
> @run driver jdk.test.lib.util.JavaAgentBuilder ...
> ```
> 
> Just like previously, since this is a test library class, jtreg tries to locate this class in the shared directory and it finds it in the shared directory `build/macosx-aarch64/test-support/jtreg_test_hotspot_jtreg_tier1_serviceability/classes/0/test/lib` (since it was compiled to that location by an unrelated test). Since it finds the class, jtreg then skips compilation of that test library class and thus the `Utils.class` (or any of the classes referenced by `JavaAgentBuilder`) isn't recompiled again. jtreg then launches the test with this shared directory and this test's specific work directory (which is different from the AgentWithVThreadTest's work directory) in the classpath. Since neither of these directories contain the `Utils.class`, we end up with this missing class error.
> 
> Like I noted, this is a known problem in jtreg. I'll see if we can do something in that area.
> 
> For now though, I think one workaround is to add a:
> 
> ```
> @clean jdk.test.lib.util.JavaAgentBuilder
> ```
> 
> before the `@run` tag to allow for the jdk.test.lib.util.JavaAgentBuilder and its referenced classes to be compiled afresh (into the shared test library directory). This is what the change would look like in your PR:
> 
> ```diff
> diff --git a/test/hotspot/jtreg/serviceability/jvmti/vthread/TestPinCaseWithCFLH/TestPinCaseWithCFLH.java b/test/hotspot/jtreg/serviceability/jvmti/vthread/TestPinCaseWithCFLH/TestPinCaseWithCFLH.java
> index 02755a0289f..60564115f51 100644
> --- a/test/hotspot/jtreg/serviceability/jvmti/vthread/TestPinCaseWithCFLH/TestPinCaseWithCFLH.java
> +++ b/test/hotspot/jtreg/serviceability/jvmti/vthread/TestPinCaseWithCFLH/TestPinCaseWithCFLH.java
> @@ -34,7 +34,7 @@
>   * @requires vm.jvmti
>   * @modules java.base/java.lang:+open
>   * @compile TestPinCaseWithCFLH.java
> - * @build jdk.test.lib.Utils
> + * @clean jdk.test.lib.util.JavaAgentBuilder
>   * @run driver jdk.test.lib.util.JavaAgentBuilder
>   *             TestPinCaseWithCFLH TestPinCaseWithCFLH.jar
>   * @run main/othervm/timeout=100  -Djdk.virtualThreadScheduler.maxPoolSize=1
> ```
> 
> I admit it's odd to be expecting the test to be doing this, but I think this should make the test stable. Can you give it a try?

Thank you for your help. I already make it by adding `@build jdk.test.lib.Utils`. Do I need to try to add `@clean jdk.test.lib.util.JavaAgentBuilder` and test again?

Reply:
I am not too sure adding the `@build jdk.test.lib.Utils` is a good thing. This test definition nor the test code uses/references that class. So it's odd to be adding a build tag for an indirect dependent class (and only that specific class). I felt the `@clean jdk.test.lib.util.JavaAgentBuilder` would be a better option since that `jdk.test.lib.util.JavaAgentBuilder` class is being used by the test definition.

Having said that, it's just a personal opinion and I would let hotspot and serviceability area members to decide what approach to use here.

Reply:
Building a test library class not actually used by the test is certainly somewhat odd. I wasn't aware of the `@clean` workaround but I see a lot of vmTestbase tests use it, so please try the `@clean` as Jai suggested. Thanks

Reply:
`@clean jdk.test.lib.util.JavaAgentBuilder` cannot solve the problem. 
I find that the reason why `AgentWithVThreadTest.java` can pass is that when compiling `AgentWithVThread.java`, it uses jdk.test.lib.Utils through `import` (`import jdk.test.lib.process.ProcessTools;`, `import jdk.test.lib.Utils;` is in the ProcessTools.java), so the test dir contains jdk.test.lib.Utils. However, my new testcase doesn't use it. 
![image](https://github.com/user-attachments/assets/cd80f76e-24b4-463b-bcfd-4b4ba32374cd)

I tryed add `import jdk.test.lib.Utils;` in my testcase, it can pass.
![image](https://github.com/user-attachments/assets/cfb10e17-23ae-47f4-b0da-e107fe06c711)


If I only run the new testcase, it can pass and the work dir look like this:
![image](https://github.com/user-attachments/assets/ade018b0-c0b0-4675-b3a0-784b4ef82b78)

If I run `AgentWithVThread.java` and then `TestPinCaseWithCFLH.java`, the Utils.class is missed in test/lib/jdk/test/lib:
![image](https://github.com/user-attachments/assets/83dd910c-ff61-4c06-9ccc-256b2d0ccf1f)

Reply:
A stable way to reproduce the problem: run AgentWithVThread.java and then TestPinCaseWithCFLH.java.

```
jtreg -v:error,fail -jdk:{JDKPATH} ./test/hotspot/jtreg/serviceability/jvmti/vthread/premain/AgentWithVThreadTest.java
jtreg -v:error,fail -jdk:{JDKPATH} ./test/hotspot/jtreg/serviceability/jvmti/vthread/TestPinCaseWithCFLH/TestPinCaseWithCFLH.java
```

Reply:
Thanks for your help. I changed it back.